data/tokenizer.py

```diff
@@ -0,0 +1,238 @@
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.tgt_bins = tgt_bins
+        if keys is None:
+            self.keys = ["pitch", "dstart_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<CLS>", "<pad>"]
+
+        self.vocab = list(self.specials)
+
+        # add midi tokens to vocab
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def __rich_repr__(self):
+        yield "MultiStartEncoder"
+        yield "vocab_size", self.vocab_size
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def _build_vocab(self):
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin key
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
+        )
+
+        for pitch in range(21, 109):
+            self.vocab.append(f"{pitch}p")
+        for start, duration in time_tokens_product:
+            self.vocab.append(f"{start:0.0f}-{duration:0.0f}")
+        for velocity in range(self.quantization_cfg.velocity):
+            self.vocab.append(f"{velocity:0.0f}v")
+        for start in range(self.tgt_bins):
+            self.vocab.append(f"{start}s")
+
+    def tokenize_src(self, record: dict) -> list[str]:
+        tokens = []
+        n_samples = len(record[self.keys[0]])
+        for idx in range(n_samples):
+            # append p and v so that there are no duplicate tokens
+            pitch_token = f"{record[self.keys[0]][idx]:0.0f}p"
+            time_token = f"{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
+            velocity_token = f"{record[self.keys[3]][idx]:0.0f}v"
+            tokens += [pitch_token, time_token, velocity_token]
+
+        return tokens
+
+    def quantize_start(self, starts: pd.Series):
+        bin_edges = np.linspace(start=0, stop=len(starts), num=self.tgt_bins)
+
+        quantized_starts = np.digitize(starts, bin_edges)
+        return quantized_starts
+
+    def tokenize_tgt(self, record: dict) -> list[str]:
+        quantized_starts = self.quantize_start(record["start"])
+        tokens = [f"{start_bin:0.0f}s" for start_bin in quantized_starts]
+        return tokens
+
+    def untokenize_src(self, tokens: list[str]) -> pd.DataFrame:
+        samples = []
+        buff = []
+        if tokens[0] == "<CLS>":
+            # get rid of cls token
+            tokens = tokens[1:]
+
+        for idx, token in enumerate(tokens):
+            if token == "<pad>":
+                break
+            if idx % 3 == 0:
+                # pitch token
+                buff.clear()
+                # [:-1] to clear p letter from a token
+                buff.append(eval(token[:-1]))
```

Comment:
```suggestion
                buff.append(int(token[:-1]))
```

I'm not sure if that's what's happening here, but using eval for string -> number conversion is not recommended, and you should use float() or int() instead